### PR TITLE
MAILBOX-351 ReIndexing only failed items of previous reIndexing tasks

### PR DESCRIPTION
**Pull request by chibenwa**
This is a request from Linagora OPS team. They agreed with the proposed migration strategy however they would like us to improve reIndexing tools.

Today if a reIndexing fails with (let's say) 283 failures an external script needs to be written to get the failed items and retry them one by one. Obviously, not ideal (one by one, external script to maintain).

The proposal is to have a endpoint for reIndexing the messages that a previous task recorded as failed.

With this tool the OPS team would build enough confidence to apply our proposed ES6 migration strategy.

### CODE REVIEW

File: src/site/markdown/server/manage-webadmin.md [hunk 1]
```diff
@@ -2493,6 +2505,48 @@ Warning: Canceling this task should be considered unsafe as it will leave the cu
...
+ - Location header indicates the location of the resource associated with the scheduled task. Example:
+
+```
+Location: /tasks/3294a976-ce63-491e-bd52-1b6f465ed7a2
+```
+
+Response codes:
+
+ - 201: Success. Corresponding task id is returned.
+ - 400: Error in the request. Details can be found in the reported error.
```
Comment (trantienduchn) [inline src/site/markdown/server/manage-webadmin.md]: s/can be/can not be

File: src/site/markdown/server/manage-webadmin.md [hunk 2]
```diff
@@ -2392,7 +2392,11 @@ The scheduled task will have the following type `FullReIndexing` and the followi
 ```
 {
   "successfullyReprocessMailCount":18,
-  "failedReprocessedMailCount": 1
+  "failedReprocessedMailCount": 3,
+  "failures": {
+    "mbx1": [{"uid": 35}, {"uid": 45}],
+    "mbx2": [{"uid": 38}],
```
Comment (trantienduchn) [inline src/site/markdown/server/manage-webadmin.md]: can you remove the comma `,` at the end? and places below

File: server/protocols/webadmin/webadmin-mailbox/src/main/java/org/apache/james/webadmin/routes/ReindexingRoutes.java [hunk 3]
```diff
@@ -110,18 +116,78 @@ public void define(Service service) {
...
+        boolean userReIndexing = !Strings.isNullOrEmpty(request.queryParams(USER_QUERY_PARAM));
+        boolean indexingCorrection = !Strings.isNullOrEmpty(request.queryParams(RE_INDEX_FAILED_MESSAGES_QUERY_PARAM));
+        if (userReIndexing && indexingCorrection) {
+            return rejectInvalidQueryParameterCombination();
+        }
+        if (userReIndexing) {
+            return wrap(request, response, () -> reIndexer.reIndex(extractUser(request)));
+        }
+        if (indexingCorrection) {
+                IndexingDetailInformation indexingDetailInformation = retrieveIndexingExecutionDetails(request);
```
Comment (Arsnael) [inline server/protocols/webadmin/webadmin-mailbox/src/main/java/org/apache/james/webadmin/routes/ReindexingRoutes.java]: indent

File: server/protocols/webadmin/webadmin-mailbox/src/main/java/org/apache/james/webadmin/service/PreviousReIndexingService.java [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+import org.apache.james.task.TaskNotFoundException;
+
+public class PreviousReIndexingService {
+    public static class TaskNotYetFinishedException extends RuntimeException {
+        TaskNotYetFinishedException(TaskManager.Status currentStatus) {
+            super("Task is not yet finished. Current status is: " + currentStatus);
+        }
+    }
+
+    public static class NotAnIndexingRetryiableTask extends RuntimeException {
```
Comment (Arsnael) [inline server/protocols/webadmin/webadmin-mailbox/src/main/java/org/apache/james/webadmin/service/PreviousReIndexingService.java]: s/Retryiable/Retriable

File: src/site/markdown/server/manage-webadmin.md (hunk 1 above)
Comment (chibenwa) [reply src/site/markdown/server/manage-webadmin.md]: Why? We do have the details in the error report...

Comment (trantienduchn) [reply src/site/markdown/server/manage-webadmin.md]: sorry, miss reading

**Comment by chibenwa** [comment]
:green_apple: before CI restart

File: mailbox/tools/indexer/src/main/java/org/apache/mailbox/tools/indexer/IndexingDetailInformation.java [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.mailbox.tools.indexer;
+
+public interface IndexingDetailInformation {
+    int getSuccessfullyReprocessMailCount();
```
Comment (remk) [inline mailbox/tools/indexer/src/main/java/org/apache/mailbox/tools/indexer/IndexingDetailInformation.java]: s/Reprocess/Reprocessed

File: mailbox/tools/indexer/src/main/java/org/apache/mailbox/tools/indexer/PreviousFailuresReIndexationTask.java [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+
+package org.apache.mailbox.tools.indexer;
+
+import java.util.Optional;
+
+import org.apache.james.mailbox.indexer.ReIndexingExecutionFailures;
+import org.apache.james.task.Task;
+import org.apache.james.task.TaskExecutionDetails;
+
+public class PreviousFailuresReIndexationTask implements Task {
```
Comment (blackheaven) [inline mailbox/tools/indexer/src/main/java/org/apache/mailbox/tools/indexer/PreviousFailuresReIndexationTask.java]: ErrorRecoveryIndexationTask

Comment (chibenwa) [reply mailbox/tools/indexer/src/main/java/org/apache/mailbox/tools/indexer/PreviousFailuresReIndexationTask.java]: Thanks for the cool name!

**Comment by chibenwa** [comment]
(just rebased on master, don't worry)

**Comment by chibenwa** [comment]
test this please

**Comment by trantienduchn** [comment]
```
[7250acbd16b353669d9b6bbbfeeef3e4ce7618a2] [ERROR]   JmapJamesServerContract.connectJMAPServerShouldRespondBadRequest » Provision U...
[7250acbd16b353669d9b6bbbfeeef3e4ce7618a2] [ERROR]   CassandraMailRepositoryIntegrationTest.deniedSenderMailShouldBeStoredInCassandraMailRepositoryWhenConfigured » Provision
[7250acbd16b353669d9b6bbbfeeef3e4ce7618a2] [ERROR]   DefaultCassandraJamesServerTest.cassandraJamesServerShouldStartWithNoConfigurationFile » Provision
[7250acbd16b353669d9b6bbbfeeef3e4ce7618a2] [ERROR]   ESReporterTest.timeMetricsShouldBeReportedWhenImapCommandsReceived » Provision
[7250acbd16b353669d9b6bbbfeeef3e4ce7618a2] [ERROR]   ESReporterTest.timeMetricsShouldBeReportedWhenJmapRequestsReceived » Provision
[7250acbd16b353669d9b6bbbfeeef3e4ce7618a2] [ERROR]   FixingGhostMailboxTest.ghostMailboxBugShouldDiscardOldContent » Provision Unab...
[7250acbd16b353669d9b6bbbfeeef3e4ce7618a2] [ERROR]   FixingGhostMailboxTest.oldGhostedMailboxShouldNoMoreBeAccessible » Provision U...
[7250acbd16b353669d9b6bbbfeeef3e4ce7618a2] [ERROR]   JamesServerWithRetryConnectionTest.serverShouldRetryToConnectToCassandraWhenStartService:90->assertThatServerStartCorrectly:114 » Provision
[7250acbd16b353669d9b6bbbfeeef3e4ce7618a2] [ERROR]   JamesServerWithRetryConnectionTest.serverShouldRetryToConnectToElasticSearchWhenStartService:99->assertThatServerStartCorrectly:114 » Provision
[7250acbd16b353669d9b6bbbfeeef3e4ce7618a2] [ERROR]   JamesServerWithRetryConnectionTest.serverShouldStartAtDefault:81->assertThatServerStartCorrectly:114 » Provision
[7250acbd16b353669d9b6bbbfeeef3e4ce7618a2] [ERROR]   CassandraSchemaVersionStartUpCheckTest.serverShouldStartSuccessfullyWhenBetweenMinAndMaxVersion:105 » Provision
[7250acbd16b353669d9b6bbbfeeef3e4ce7618a2] [ERROR]   CassandraSchemaVersionStartUpCheckTest.serverShouldStartSuccessfullyWhenMaxVersion:95 » Provision
[7250acbd16b353669d9b6bbbfeeef3e4ce7618a2] [ERROR]   CassandraSchemaVersionStartUpCheckTest.serverShouldStartSuccessfullyWhenMinVersion:115 » Provision
```

**Comment by chibenwa** [comment]
```
[7250acbd16b353669d9b6bbbfeeef3e4ce7618a2] 1) Error in custom provider, java.lang.IllegalStateException: Retries exhausted: 7/7
[7250acbd16b353669d9b6bbbfeeef3e4ce7618a2]   at org.apache.james.modules.mailbox.ElasticSearchMailboxModule.provideClient(ElasticSearchMailboxModule.java:140) (via modules: com.google.inject.util.Modules$OverrideModule -> com.google.inject.util.Modules$OverrideModule -> com.google.inject.util.Modules$OverrideModule -> com.google.inject.util.Modules$OverrideModule -> com.google.inject.util.Modules$OverrideModule -> com.google.inject.util.Modules$OverrideModule -> com.google.inject.util.Modules$CombinedModule -> com.google.inject.util.Modules$CombinedModule -> com.google.inject.util.Modules$CombinedModule -> org.apache.james.modules.mailbox.ElasticSearchMailboxModule)
[7250acbd16b353669d9b6bbbfeeef3e4ce7618a2]   while locating org.elasticsearch.client.RestHighLevelClient
[...]
[7250acbd16b353669d9b6bbbfeeef3e4ce7618a2] 	Caused by: org.elasticsearch.client.ResponseException: method [PUT], host [http://172.17.0.1:56527], URI [/mailbox_v1?master_timeout=30s&include_type_name=false&timeout=30s], status line [HTTP/1.1 403 Forbidden]
[7250acbd16b353669d9b6bbbfeeef3e4ce7618a2] {"error":{"root_cause":[{"type":"cluster_block_exception","reason":"blocked by: [FORBIDDEN/12/index read-only / allow delete (api)];"}],"type":"cluster_block_exception","reason":"blocked by: [FORBIDDEN/12/index read-only / allow delete (api)];"},"status":403}
```

test this please

**Comment by trantienduchn** [comment]
```
[7250acbd16b353669d9b6bbbfeeef3e4ce7618a2] Location: /org/apache/james/imap/scripts/ValidateDeployment.test:499
[7250acbd16b353669d9b6bbbfeeef3e4ce7618a2] LastClientMsg: c SEARCH FLAGGED
[7250acbd16b353669d9b6bbbfeeef3e4ce7618a2] Expected: '\* SEARCH 1 2 3 4'
[7250acbd16b353669d9b6bbbfeeef3e4ce7618a2] Actual   : '* SEARCH'
```

test this please

**Comment by chibenwa** [comment]
test this please

**Comment by chibenwa** [comment]
Merged

**Pull request closed without merging**